src/acceptance-test/java/org/zalando/nakadi/webservice/hila/HilaRebalanceAT.java

```diff
@@ -125,6 +128,132 @@ public void whenRebalanceThenPartitionsAreEquallyDistributedAndCommittedOffsetsA
         waitFor(() -> assertThat(clientA.getBatches(), hasSize(100)));
     }
 
+    @Test(timeout = 15000)
+    public void whenDirectlyRequestPartitionsTheyAssignedCorrectly() throws IOException {
+        // publish 2 events to each partition
+        publishBusinessEventWithUserDefinedPartition(
+                eventType.getName(), 16, x -> "blah" + x, x -> String.valueOf(x % 8));
+
+        // start a stream requesting to read from partitions 5, 6
+        final TestStreamingClient client = new TestStreamingClient(URL, subscription.getId(), "", Optional.empty(),
+                Optional.of("{\"partitions\":[" +
+                        "{\"event_type\":\"" + eventType.getName() + "\",\"partition\":\"6\"}," +
+                        "{\"event_type\":\"" + eventType.getName() + "\",\"partition\":\"5\"}" +
+                        "]}"));
+        client.start();
+
+        // wait till we receive 4 batches (2 per partition)
+        waitFor(() -> assertThat(client.getBatches(), hasSize(4)));
+
+        // check that all batches are from partitions 5, 6
+        checkAllEventsAreFromPartitions(client, ImmutableSet.of("5", "6"));
+    }
+
+    @Test(timeout = 15000)
+    public void whenMixedStreamsThenPartitionsAssignedCorrectly() throws IOException, InterruptedException {
+
+        // start 2 streams not specifying partitions directly
+        final TestStreamingClient autoClient1 = new TestStreamingClient(URL, subscription.getId(), "");
+        autoClient1.start();
+        final TestStreamingClient autoClient2 = new TestStreamingClient(URL, subscription.getId(), "");
+        autoClient2.start();
+
+        // start a stream requesting to read from partition 6
+        final TestStreamingClient directClient1 = new TestStreamingClient(URL, subscription.getId(), "",
+                Optional.empty(),
+                Optional.of("{\"partitions\":[" +
+                        "{\"event_type\":\"" + eventType.getName() + "\",\"partition\":\"6\"}]}"));
+        directClient1.start();
+
+        // start a stream requesting to read from partition 7
+        final TestStreamingClient directClient2 = new TestStreamingClient(URL, subscription.getId(), "",
+                Optional.empty(),
+                Optional.of("{\"partitions\":[" +
+                        "{\"event_type\":\"" + eventType.getName() + "\",\"partition\":\"7\"}]}"));
+        directClient2.start();
+
+        // wait for rebalance to finish
+        Thread.sleep(1000);
```

Comment:
This is a bit brittle, but I don't really see a way to waitFor() the end of the rebalance (as opposed to waiting for some time, hoping that the rebalance will be finished)

Reply:
I think it's possible to check that using stats endpoint. I will check if I can use that.

Reply:
Done https://github.com/zalando/nakadi/pull/848/commits/ec3f4429aabf2181e0f1ffbb5f5643dce9fe9970